docs/src/design/changing_the_primal.md

```diff
@@ -0,0 +1,435 @@
+# Design Notes: Why can you change the primal computation?
+#TODO generalize intro/title to ask why `rrule` is how it is, and in particular about changing primal
+
+These design notes are to help you understand why ChainRules allows the primal computation to be changed.
+We will focus this discussion on reverse mode and `rrule`, though the same also applies to forwards mode and `frule`.
+In fact, it has a particular use in forward mode for efficiently calculating the pushforward of a differential equation solve via expanding the system of equations to also include the derivatives, and solving all at once.
+In forward mode it is related to the fusing of `frule` and `pushforward`.
```

Comment:
for primal `y=f(x)` we have a pushfward given by  is: ` ẏ=pushforward(ẋ)
It can be defused out of `frule`
via:
```julia
function linearize(f, x...)
    function pushfoward(ẋ...)
        y, ẏ = frule(ẋ, f, x...)
        return ẏ
    end
    return pushfoward
end
```

Now if the optimizer runs well-enough, we would be able to be JAXish.

The defination of the pushforward is given in the main page of the docs.
Since this is kind of dev-docs, I don't think we need to give it again?